### PR TITLE
Add profile url search

### LINKED ISSUE #7984: Search a unknown profile URL should probe and import the contact
The same way we probe and import public post from their URL in the search, we should try to probe for contact profiles and import them.

### PR DESCRIPTION
**Pull request by MrPetovan**
Closes #7984 

This PR adds the profile URL search like we do with post URLs: it redirects to the contact page if it found one at the URL.

I separated the anonymous search from the logged in search to avoid contact probes by anonymous users. @annando can you please check to make sure I used the right `Model\Contact` method for this purpose?

Corollary question: Should I separate the post URL search between anonymous and logged in users?

Or third option, should those searches be logged-users only?

### CODE REVIEW

**Comment by annando** [comment]
From an UX standpoint I suggest that both searches - this new one and the existing one with `@` should behave identically.

**Comment by MrPetovan** [comment]
Unfortunately it can't. The `@` and `!` local search matches the `url`, `name`, `location`, `addr`, `about` and `keywords` fields in the `gcontact` table and not all of them are unique. The global search hits the global directory endpoint `/search` which can return multiple results as well depending on the input.

**Comment by annando** [comment]
I see the complicity of the situation. But I do see the user experience as well. Because of tis I would like to have the same user experience for a single contact for both ways of searching.

**Comment by MrPetovan** [comment]
I disagree, Entering a post URL or a profile URL means you want the related uniquely identified object to be available on your local node, while a free text search may yield multiple results and needs a different behavior with a result list.

**Comment by annando** [comment]
Have a look at this code: https://github.com/friendica/friendica/blob/develop/src/Module/BaseSearchModule.php#L56-L58

The function `Search::getContactsFromProbe` is called only in the special case when the search pattern looks like a complete contact.

So I would vote for either call `Search::getContactsFromProbe` in any case (even the above one). Or in the case of a search pattern like `@user@host.tld` or `@http://host.tld/.../user` we should call the new functionality.

**Comment by MrPetovan** [comment]
I'd rather go for the latter. The `performContactSearch` is done when we know we have a contact-like query (starting with `@` or `!`) but a given URL may not be a profile URL.

Additionally, can you please answer my question about anonymous users having access to the profile and post URL search?

**Comment by annando** [comment]
I have to have a look at this anonymous search. But this will be done this evening.

**Comment by MrPetovan** [comment]
I moved both profile and post URL searches to their own methods for clarity. The profile URL search now includes webfinger addresses as well.

File: src/Module/Search/Index.php [hunk 1]
```diff
@@ -197,4 +196,78 @@ public static function content(array $parameters = [])
...
+	 *
+	 * @param BaseURL $baseURL
+	 * @param string  $search
+	 * @throws HTTPException\InternalServerErrorException
+	 * @throws \ImagickException
+	 */
+	private static function tryRedirectToProfile(BaseURL $baseURL, string $search)
+	{
+		if (local_user() && preg_match('/^@?([a-z0-9.-_]+@[a-z0-9.-_:]+)$/i', trim($search), $matches)) {
+			$contact = Contact::getDetailsByAddr($matches[1]);
```
Comment (annando) [inline src/Module/Search/Index.php]: You can use `Contact::getIdForURL` for addresses as well.

**Comment by MrPetovan** [comment]
Reworked again the method, now t also looks up webfinger addresses for anonymous users.

File: src/Module/Search/Index.php [hunk 2]
```diff
@@ -197,4 +196,89 @@ public static function content(array $parameters = [])
...
+	/**
+	 * Fetch/search a post by URL and redirects to its local representation if it was found.
+	 *
+	 * @param BaseURL $baseURL
+	 * @param string  $search
+	 * @throws HTTPException\InternalServerErrorException
+	 */
+	private static function tryRedirectToPost(BaseURL $baseURL, string $search)
+	{
+		if (parse_url($search, PHP_URL_SCHEME) != '') {
```
Comment (annando) [inline src/Module/Search/Index.php]: I would prefer exiting with `return` at the top of the function when the check isn't successful. I dislike large `if` constructs that reach through the whole function.

Comment (MrPetovan) [reply src/Module/Search/Index.php]: Fair enough.

**Comment by MrPetovan** [comment]
One less indentation level!

**Pull request merged**